File: engine/src/pptx_mcp/schema.py
from dataclasses import asdict

from pptx import Presentation

from .assembler import find_shape
from .autodetect import DEFAULT_FONT_PT, _first_font_pt, estimate_max_chars
from .models import Template


def _slot_dict(slot) -> dict:
    return {
        "id": slot.id, "name": slot.name, "type": slot.type,
        "required": slot.required, "default": slot.default,
        "constraints": {k: v for k, v in asdict(slot.constraints).items() if v is not None},
    }


def _slot_geometry(slide, slot) -> dict | None:
    try:
        shape = find_shape(slide, slot.shape_id)
    except KeyError:
        return None
    w, h = int(shape.width or 0), int(shape.height or 0)
    font_pt = capacity = None
    if slot.type == "text":
        font_pt = _first_font_pt(shape) or DEFAULT_FONT_PT
        capacity, _ = estimate_max_chars(w, h, font_pt)
    return {"width_emu": w, "height_emu": h,
            "font_pt": font_pt, "capacity_chars": capacity}


def get_schema(template: Template) -> dict:
    prs = Presentation(template.pptx_path)
    slide_types = []
    for st in template.slide_types:
        slide = prs.slides[st.source_slide_index]
        slots = [{**_slot_dict(s), "geometry": _slot_geometry(slide, s)} for s in st.slots]
        slide_types.append({
            "id": st.id, "name": st.name, "description": st.description,
            "slots": slots,
        })
    return {
        "id": template.id, "name": template.name, "description": template.description,
        "slide_types": slide_types,
    }

File: web/src/lib/schema.ts
import { buildExampleDeckSpec } from "@/lib/example";

export function toAgentSchema(manifestJson: any, meta?: { id: string; name: string; description: string }) {
  return {
    id: meta ? meta.id : manifestJson?.template?.id,
    name: meta ? meta.name : manifestJson?.template?.name,
    description: meta ? meta.description : (manifestJson?.template?.description ?? ""),
    slide_types: (manifestJson?.slide_types ?? []).map((st: any) => ({
      id: st.id, name: st.name, description: st.description ?? "",
      repeatable: st.repeatable ?? false,
      slots: (st.slots ?? []).map((s: any) => ({
        id: s.id, name: s.name, type: s.type,
        required: s.required ?? true, default: s.default ?? null,
        constraints: s.constraints ?? {},
        description: s.description ?? "",
        example: s.example ?? null,
      })),
    })),
    example_deck_spec: buildExampleDeckSpec(manifestJson),
  };
}
